text/0813-template-layering.md

```diff
@@ -0,0 +1,856 @@
+this does not extend to the "tagged template string" format. Attempting to use
+the `template()` function as a tag at runtime will result in a helpful error in
+debug builds and undefined behavior on production builds.
+
+If the template compiler is not available at runtime, then attempts to call the
+`template()` function at runtime will result in a helpful error in debug builds
+and undefined behavior on production builds. On production builds, it is not
+guaranteed that the `template` named export may be undefined (i.e. it may not
+even be a function), but developers must not rely on this to "feature-detect"
+the availability of runtime compilation.
+
+Notably, even when template compilation is available at runtime, the result of
+the compilations may be subtly different. This is because that applications may
+have custom glimmer/handlebars AST plugins in their build, and these plugins
+will not be available at runtime.
+
+This may change in the future, but it has always been true for runtime template
+compilation and will likely remain to be the case for the foreseeable future,
+as these plugins are often authored or packaged in ways that are inherently not
+browser-friendly. This usually only used for optional syntax extensions such as
+"inline {{let}}" so in practice it is not an issue. However, some polyfills for
+template features also uses the AST plugin API and those polyfilled features
+will not be available to templates compiled at runtime.
+
+#### Opting In
+
+Today, applications can opt-in to bundling the runtime compiler with something
+along these lines in their `ember-cli-build.js`:
+
+```js
+app.import("node_modules/ember-source/dist/ember-template-compiler.js");
+```
+
+The exact details depend on the version of `ember-source` and is not very well
+documented. It also does not align with the direction we are headed, where we
+encourage using the modules system and imports to express dependencies. This
+configuration is also not friendly to tree-shaking/code-splitting, as it forces
+the template compiler to be included into the main/initial bundle, when it may
+only be needed in one of the infrequently used routes.
+
+This RFC takes the opportunity to improve on this by proposing a new module –
+`@ember/template-compilation/runtime`. When this module is imported, either as
+a "side-effect" import statement, or if its exported values are imported, then
+it serves as an opt-in for enabling runtime template compilation.
+
+For the time being, we propose that this module should have a single named
+export – `template`, which is a re-export of the `template()` function we
+described above. It has the same semantics as `template()` from the main module
+except it is always invalid to use this imported function as a tag, so linters
+and TypeScript can trivial flag this invalid usage.
+
+The purpose of the re-exported `template()` from the runtime module is to
```

Comment:
will we throw errors if the non-runtime module makes it into a build? (even if accidentally somehow, where a linter wouldn't be able to help?)

or, does it matter?

If the non-runtime module _happens_ to make it into the build, is it the exact same as the runtime module?

what are the drawbacks of only having one module instead of also having the runtime module?

Reply:
It won't. `template()` from the runtime module is just a re-export so yeah they are exactly the same. The reason to import from the main module (plus sticking with the statically analyzable subset, both of which you should do 95% of the time), is for your code to be agnostic to where the compilation happens. The reason to import from the runtime module is to make it clear that the template should never be processed at build time.

Reply:
so is it just a hint to tooling? does it instead make sense to provide a way to analyze usage of `template` within a module to determine if it should be left in or compiled away?

Reply:
Sure. It's more than a guarantee than a hint, and it's not just to tools, it also makes it clear to whoever is reading the code that it is deliberate. You can look at the "idiomatic usages" section. The idea is we should implement all the same rules in the linter and warn you about non-idiomatic usages. And the instructions will tell you to either not do the thing that made it not statically analyzable, or to switch over to use the runtime import to make it clear that it's deliberate. However it's "just a lint". Even if you don't change the import and let it ship to the browser, it would still work (but you will probably get an additional warning).

Reply:
I think I cleared this up some more in the most recent update